valhalla/meili/match_result.h

```diff
@@ -26,25 +26,62 @@ struct MatchResult {
   double epoch_time;
   // Sequential state id
   StateId stateid;
+  // Whether or not this was a break or break_through type location
+  bool is_break_point;
+  // Whether or not there is a discontinuity starting from this point
+  bool begins_discontinuity;
+  // Whether or not a discontinuity ends at this point
+  bool ends_discontinuity;
+  // An index into the full list of edges in the path even across discontinuities (for
+  // trace_attributes)
+  size_t edge_index;
 
   bool HasState() const {
     return stateid.IsValid();
   }
+
+  enum class Type { kUnmatched, kInterpolated, kMatched };
+  Type GetType() const {
+    // Set the type based on edge id and state
+    if (edgeid.Is_Valid() && HasState()) {
+      return Type::kMatched;
+    } else if (edgeid.Is_Valid()) {
+      return Type::kInterpolated;
+    } else {
+      return Type::kUnmatched;
+    }
+  }
 };
 
 struct EdgeSegment {
-  EdgeSegment(baldr::GraphId the_edgeid, float the_source = 0.f, float the_target = 1.f);
+  EdgeSegment(baldr::GraphId the_edgeid,
+              float the_source = 0.f,
+              float the_target = 1.f,
+              int the_first_match_idx = -1,
+              int the_last_match_idx = -1,
+              bool disconnect = false);
 
   std::vector<midgard::PointLL> Shape(baldr::GraphReader& graphreader) const;
 
   bool Adjoined(baldr::GraphReader& graphreader, const EdgeSegment& other) const;
 
   // TODO make them private
   baldr::GraphId edgeid;
-
-  float source;
-
-  float target;
+  float source{0.f};
+  float target{1.f};
+  int first_match_idx{-1};
+  int last_match_idx{-1};
+  bool discontinuity{false};
+
+  // Stream output
+  friend std::ostream& operator<<(std::ostream& os, const EdgeSegment& segment) {
```

Comment:
nit: Perhaps @yuzheyan can clarify here: does this need to be a friend member function or can it be a bare function? My understanding is that it can be the latter.

Reply:
doesnt need to be but there is no benefit to having it be a free function vs a friend. organizationally i prefer it as is